working/control-flow-collections/implementation-plan.md

```diff
@@ -0,0 +1,189 @@
+# Implementation Plan for "Control Flow Collections"
+
+Owner: rnystrom@google.com ([@munificent](https://github.com/munificent/) on GitHub)
+
+Relevant links:
+
+* [Tracking issue](https://github.com/dart-lang/language/issues/78)
+* [Proposal](https://github.com/dart-lang/language/blob/master/working/control-flow-collections/feature-specification.md)
+
+## Phase 0 (Prerequisite)
+
+### "control-flow-collections" Experimental flag
+
+The implementation of this feature should be hidden behind an [experiment
+flag][]. Tools must be passed the flag
+`--enable-experiment=control-flow-collections` to enable the feature.
+
+[experiment flag]: https://github.com/dart-lang/sdk/blob/master/docs/process/experimental-flags.md
+
+While this feature is under development, individual tools may have incomplete or
+changing implementations behind the flag. When all tools have completely
+implemented the feature, the the feature will be enabled by default, and the
+flag removed in a stable release.
+
+### Tests
+
+The language team adds tests for the feature.
+
+## Phase 1 (Foundation)
+
+### CFE
```

Comment:
We will plan to have constant evaluation in the CFE to support constant control-flow collections.  We don't have that yet, so that is a prerequisite.  CFE team will have to make that a very high priority.

Our implementation will involve a 'block expression' in Kernel which is new and first introduced for this purpose.  So depending on how you look at it, the back ends will need to support both a constant pool in .dill files and block expressions.

I'll work on estimates for those tasks.

Reply:
Thanks, added a note.